Update slotmap when slot is not served by any node

bjosv opened the pull request:
The sync-API will now update the slotmap before attempting to send a command when the slot is not served.
    
The async-API will initate the slotmap update in parallell so that the next command might have an updated slotmap.

This also fix an additional issue by using the correct function name for `redisClusterUpdateSlotmap()`.
The former name `cluster_update_route()` is not defined when `HIRCLUSTER_NO_OLD_NAMES` is defined, which would break the build.

Fixes #191

zuiderkwast commented on tests/ct_connection.c:
```diff
@@ -401,6 +401,9 @@ void test_async_password_wrong(void) {
     assert(acc->err == REDIS_ERR_OTHER);
     assert(strcmp(acc->errstr, "slotmap not available") == 0);
 
+    /* Since the owner of the slot is unknown the redisClusterAsyncCommand will
+       initite a slotmap update. This update will also get a WRONGPASS error. */
+
     event_base_dispatch(base);
```
... and after WRONGPASS, it disconnects which removes the fd from the event loop which exits?

I am slightly confused about what happens behind the scenes.

bjosv replied:
Reverted this change. There is no need for an eventloop to test the WRONGPASS scenario here.

zuiderkwast commented on tests/ct_connection.c:
```diff
@@ -434,7 +437,12 @@ void test_async_password_missing(void) {
     assert(strcmp(acc->errstr, "slotmap not available") == 0);
 
-    event_base_dispatch(base);
+    /* Since the owner of the slot is unknown the redisClusterAsyncCommand will
+       initite a slotmap update. This update will register a new socket fd in the
+       event system, which will block libevents API `event_base_dispatch()` until
+       the socket is closed. Use alternative API to run the event loop once only
+       to avoid this problem. */
+    event_base_loop(base, EVLOOP_ONCE);
 
     redisClusterAsyncFree(acc);
```
Why run the event loop at all then? Which event is it handling and shouldn't we check afterwards that it did what we expected?

bjosv replied:
Good idea, I have removed the use of an eventloop in this case.

zuiderkwast commented on tests/scripts/slots-not-served-test-async.sh:
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+# Usage: $0 /path/to/clusterclient-binary
```
```suggestion
# Usage: $0 /path/to/clusterclient-async
```

This test needs the async client, so better make it clear in this comment too.

zuiderkwast reviewed (commented):
Very nice!

Only some questions about the other (slightly unrelated) fixes.

bjosv commented on tests/clusterclient_async.c:
```diff
@@ -157,8 +157,15 @@ void sendNextCommand(int fd, short kind, void *arg) {
         } else {
             int status = redisClusterAsyncCommand(
                 acc, replyCallback, (void *)((intptr_t)num_running), cmd);
-            ASSERT_MSG(status == REDIS_OK, acc->errstr);
-            num_running++;
+            if (status == REDIS_OK) {
+                num_running++;
+            } else {
+                printf("error: %s\n", acc->errstr);
```
We just didn't test it. All previous testcases are able to put the command on hiredis internal queue, but then might get errors via the callback as NULL replies.

bjosv commented:
I reverted the callback handling for AUTH commands, to be added via separate PR.
There are other authentication related changes that can be discussed in that PR.

Also removed the use of the eventloop in the testcases that didn't need any events for its testing.

zuiderkwast reviewed (approved):
Looks good!

Let's take the other things in other PRs.
